sys/riscv/trap.c

```diff
@@ -0,0 +1,301 @@
+  return _REG(ctx, CAUSE) & SCAUSE_INTR;
+}
+
+__no_profile static inline unsigned ctx_code(ctx_t *ctx) {
+  return _REG(ctx, CAUSE) & SCAUSE_CODE;
+}
+
+__no_profile static inline unsigned ctx_intr_enabled(ctx_t *ctx) {
+  return _REG(ctx, SR) & SSTATUS_SPIE;
+}
+
+static __noreturn void kernel_oops(ctx_t *ctx) {
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+
+  klog("%s at %p!", exceptions[code], epc);
+
+  switch (code) {
+    case SCAUSE_INST_MISALIGNED:
+    case SCAUSE_INST_ACCESS_FAULT:
+    case SCAUSE_LOAD_MISALIGNED:
+    case SCAUSE_LOAD_ACCESS_FAULT:
+    case SCAUSE_STORE_MISALIGNED:
+    case SCAUSE_STORE_ACCESS_FAULT:
+    case SCAUSE_INST_PAGE_FAULT:
+    case SCAUSE_LOAD_PAGE_FAULT:
+    case SCAUSE_STORE_PAGE_FAULT:
+      klog("Caused by reference to %lx!", _REG(ctx, TVAL));
+      break;
+
+    case SCAUSE_ILLEGAL_INSTRUCTION: {
+      uint32_t badinstr = *(uint32_t *)epc;
+      klog("Illegal instruction %08x in kernel mode!", badinstr);
+      break;
+    }
+
+    case SCAUSE_BREAKPOINT:
+      klog("No debbuger in kernel!");
+      break;
+  }
+  klog("HINT: Type 'info line *%p' into gdb to find faulty code line", epc);
+
+  panic("KERNEL PANIC!!!");
+}
+
+static void page_fault_handler(ctx_t *ctx) {
+  thread_t *td = thread_self();
+
+  unsigned code = ctx_code(ctx);
+  void *epc = (void *)_REG(ctx, PC);
+  vaddr_t vaddr = _REG(ctx, TVAL);
+
+  klog("%s at %p, caused by reference to %lx!", exceptions[code], epc, vaddr);
+
+  pmap_t *pmap = pmap_lookup(vaddr);
+  if (!pmap) {
+    klog("No physical map defined for %lx address!", vaddr);
+    goto fault;
+  }
+
+  vm_prot_t access;
+  if (code == SCAUSE_INST_PAGE_FAULT)
+    access = VM_PROT_EXEC;
+  else if (code == SCAUSE_LOAD_PAGE_FAULT)
+    access = VM_PROT_READ;
+  else
+    access = VM_PROT_WRITE;
+
+  int error = pmap_emulate_bits(pmap, vaddr, access);
+  if (!error)
+    return;
+
+  if (error == EACCES || error == EINVAL)
+    goto fault;
+
+  vm_map_t *vmap = vm_map_lookup(vaddr);
+  if (!vmap) {
+    klog("No virtual address space defined for %lx!", vaddr);
+    goto fault;
+  }
+
+  if (!vm_page_fault(vmap, vaddr, access))
+    return;
+
+fault:
+  if (td->td_onfault) {
+    /* Handle copyin/copyout faults. */
+    _REG(ctx, PC) = td->td_onfault;
+  } else if (user_mode_p(ctx)) {
+    /* Send a segmentation fault signal to the user program. */
+    sig_trap(ctx, SIGSEGV);
+  } else {
+    /* Panic when kernel-mode thread uses wrong pointer. */
+    kernel_oops(ctx);
+  }
+}
+
+/*
+ * RISC-V syscall ABI:
+ *  - a7: code
+ *  - a0-5: args
+ *
+ * NOTE: the following code assumes all arguments to syscalls are passed
+ * via registers.
+ */
+static_assert(SYS_MAXSYSARGS <= FUNC_MAXREGARGS - 1,
+              "Syscall args don't fit in registers!");
+
+static void syscall_handler(mcontext_t *uctx, syscall_result_t *result) {
+  /* TODO: eventually we should have a platform-independent syscall handler. */
```

Comment:
Could you explain that comment?

Reply:
This comment is here because I saw it in the corresponding function for MIPS. I didn't ask @cahirwpz to verify if this is still meaningful but I understand that it's not. The comment has been removed.